### DISCUSSION TITLE
Network access from a VM

### DISCUSSION THREAD

**Discussion by ghost** [cloud-hypervisor/cloud-hypervisor#5196] [Q&A]
My goal is to provide network access from a VM.

I created `tap0` device on host machine:
`sudo ip tuntap add dev tap0 mode tap`
`sudo ip link set dev tap0 up`
`sudo ip addr add 192.168.1.1/24 dev tap0`

Created VM:
```
./cloud-hypervisor \
	--kernel ./hypervisor-fw \
	--disk path=focal-server-cloudimg-amd64.raw path=/tmp/ubuntu-cloudinit.img \
	--cpus boot=4 \
	--memory size=1024M \
	--net "tap=tap0,mac=00:11:22:33:44:55,ip=192.168.1.100,mask=255.255.255.0"
```

`ip addr` inside the VM:
```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: ens4: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 00:11:22:33:44:55 brd ff:ff:ff:ff:ff:ff
    inet6 fe80::211:22ff:fe33:4455/64 scope link 
       valid_lft forever preferred_lft forever
```

No internet access.

What mistake I made?
Not sure how networking works here, any advice is appreciated.

**Comment by ghost** [comment]
Also tried from: https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/docs/macvtap-bridge.md

```
sudo ./cloud-hypervisor \
	--kernel ./hypervisor-fw \
	--disk path=focal-server-cloudimg-amd64.raw path=/tmp/ubuntu-cloudinit.img \
	--cpus boot=4 \
	--memory size=2048M \
	--net fd=3,mac=$mac 3<>$"$tapdevice" \
```

Getting error:
```
Error booting VM: VmBoot(DeviceManager(CreateVirtioNet(TapError(IoctlError(2147767506, Os { code: 25, kind: Uncategorized, message: "Not a tty" })))))
```

**Reply by likebreath** [reply]
Your guest network interface (`ens4`) does not have a IP. You may need to manually assign one to it. Also, you will need to make sure your host has the right routing and iptable rules (e.g. allowing forwarding packages).

Are you using a regular terminal to run the Cloud Hypervisor command with macvtap? Also, you can run the two macvtap integration tests locally to see if macvtap works in your environment. It only requires docker to be installed. You can run them with a single command:

```bash
$ ./scripts/dev_cli.sh tests --integration -- --test-filter test_macvtap
```

**Reply by adamliyi** [reply]
@kochia5 , please refer to: https://github.com/cloud-hypervisor/cloud-hypervisor/issues/4917#issuecomment-1442834740
In this case, if starting cloud hypervisor as 'sudo', there might be ioctl error. Try login in as root, and removing sudo.

**Comment by peng6662001** [comment]
@kochia5 Pleae reference [Virtual-machine-network-configuration](https://github.com/AmpereComputing/ampere-lts-kernel/wiki/Virtual-machine-network-configuration)
